src/tools/simeng/gdb/GDBStub.cc

```diff
@@ -0,0 +1,375 @@
+#include "simeng/GDBStub.hh"
+
+namespace simeng {
+
+GDBStub::GDBStub(simeng::Core& core, simeng::MemoryInterface& dataMemory)
+    : core_(core), dataMemory_(dataMemory) {}
+
+int GDBStub::run() {
+  // verbose_ = true; // uncomment this if you'd like the GDBStub to run
+  // verbosely
+
+  int connection = openSocket(2424);  // change this if the port is blocked
+
+  std::cout << "[SimEng:GDBStub] Connection to GDB client established, "
+               "debugging in progress"
+            << std::endl
+            << std::endl;
+
+  char buffer[10000];
+
+  while (true) {
+    read(connection, buffer, 10000);
+
+    // First parentheses = package, second = checksum
+    std::regex packet_regex("^\\+*\\$([^#]*)#([0-9a-f]{2})");
+    std::smatch packet_match;
+    std::string bufferString = buffer;
+
+    if (!noAckMode_) {
+      // '+' is an acknowledgement of successful receipt of message
+      // TODO: handle '-', no acknowledgement
+      std::regex ack_regex("^\\+.*");
+      if (regex_match(bufferString, ack_regex)) {
+        if (verbose_)
+          std::cout << CYAN
+                    << "[SimEng:GDBStub] <- Received message acknowledgement"
+                    << std::endl
+                    << RESET;
+      }
+    }
+
+    if (regex_match(bufferString, packet_match, packet_regex)) {
+      std::string packet = packet_match[1].str();
+      if (verbose_) {
+        std::cout << CYAN;
+        if (packet_match.size() == 3)
+          std::cout << "[SimEng:GDBStub] <- Received packet: " << packet
+                    << std::endl;
+        else
+          std::cout << RED << "[SimEng:GDBStub] <- Unknown message"
+                    << std::endl;  // TODO: take action
```

Comment:
If there is an error then the `Unknown Message` output should arguably be displayed whether verbose mode is enabled or disabled. (Unless it is common for `unknown message` errors?)

Reply:
I believe an ``unknown message`` should only occur if something changes or is wrong on the GDB side of things, as communication over TCP should receive the messages correctly.  This check is only a simple check on the message format.  I would be happy to remove this whole ``else`` statement if you are, or have it changed to an error through SimEng's ``ExceptionHandler``